### DISCUSSION TITLE
Foundry Fundamentals/Interact with a smart contract using the CLI

### DISCUSSION THREAD

**Discussion by 0xhung295** [Cyfrin/foundry-full-course-cu#2858] [General]
call() function

**Comment by 0xhung295** [comment]
I tried to run store() function with input 77:
![image](https://github.com/user-attachments/assets/46cdd694-759c-44eb-bbcf-eeb8886b9c3f)
& then after I wanna call retrieve() function to show Favorite_Number but the result returned "0x"
![image](https://github.com/user-attachments/assets/1841c5e8-c79e-4806-aad4-9e36c0501392)

Anyone knows What's the problem in this case? Thanks

**Reply by EngrPips** [reply]
Hello @0xhung295, Please specify the `--rpc-url` when you use the `cast call` command too.

**Reply by 0xhung295** [reply]
@EngrPips : The return is the same
![image](https://github.com/user-attachments/assets/809e83f7-cbd8-481a-b587-6cad3e9855c0)

**Reply by EngrPips** [reply]
Send the address of the contract you are interacting with.

**Reply by LordAdhaar** [reply]
He's most probably working on anvil

**Reply by EngrPips** [reply]
hmm 🤔

**Reply by EngrPips** [reply]
Interesting heads-up. @0xhung295 Make sure to use the `--broadcast` flag when you run `cast send` command, and let's see if the issue persists.

**Reply by LordAdhaar** [reply]
> Interesting heads-up. @0xhung295 Make sure to use the `--broadcast` flag when you run `cast send` command, and let's see if the issue persists.

I deleted that comment, most probably it is wrong

**Reply by LordAdhaar** [reply]
Yeah, I was wrong for that comment, can't use broadcast wiht cast

**Reply by EngrPips** [reply]
I see.

**Reply by Bernz868** [reply]
I was having this similar issue doing this exercise. When I looked at that number I got the impression that nothing was recorded. I think it was a matter of running anvil and deploying the contract before you do these commands. I did that and was able to get a hex number closer to what was shown in the video.

**Reply by EngrPips** [reply]
Yeah, You are expected to deploy a contract to the blockchain before you attempt to interact with it.

**Comment by LordAdhaar** [comment]
I ran the code and successfully replicated your issue. Here's what I found:

**The Problem**
It seems you're trying to use `cast send` and `cast call` without first deploying the contract on the anvil blockchain.

**Solution**
Follow these steps in order:

1. **Open anvil**
   - Open a new terminal
   - Start anvil

2. **Deploy the contract**
   - Open another terminal
   - Run:
   ```bash
   forge script script/DeploySimpleStorage.s.sol:DeploySimpleStorage --rpc-url http://127.0.0.1:8545 --private-key $PRIVATE_KEY --broadcast
   ```

3. **Interact with the contract**
   - Store a value:
   ```bash
   cast send 0xbdEd0D2bf404bdcBa897a74E6657f1f12e5C6fb6 "store(uint256)" 1337 --rpc-url http://127.0.0.1:8545 --private-key $PRIVATE_KEY
   ```
   - Retrieve the value:
   ```bash
   cast call 0xbdEd0D2bf404bdcBa897a74E6657f1f12e5C6fb6 "retrieve()"
   ```

**Note**: Make sure to replace `0xbdEd0D2bf404bdcBa897a74E6657f1f12e5C6fb6` with your actual contract address.

**Comment by 0xhung295** [comment]
@0xhung295 @EngrPips  I followed to the solution @LordAdhaar suggested and It worked. Thanks everyone!

**Reply by LordAdhaar** [reply]
Happy to help :) Can you mark this as answered

**Reply by EngrPips** [reply]
It can't be marked as answered as it is just a discussion and not `question and answer`

**Reply by LordAdhaar** [reply]
Interesting

**Reply by LordAdhaar** [reply]
Whats the difference?

**Reply by EngrPips** [reply]
> Interesting

Indeed

**Reply by EngrPips** [reply]
> Whats the difference?

`Discussion` is meant to discuss opinion while `questions and answer` is meant to find an answer to a question. This is what makes sense to me.

**Reply by LordAdhaar** [reply]
Indeed

**Reply by EngrPips** [reply]
Yeah
